lib/board.js

```diff
@@ -1,138 +1,58 @@
 'use strict';
 
-var com = require('serialport');
-var when = require('when');
-var nodefn = require('when/node');
-
+var SerialPort = require('serialport').SerialPort;
 var bs2 = require('bs2-programmer');
+var bluebird = require('bluebird');
 
-function Board(){
-}
+var serial = require('./serial');
 
-Board.prototype.bootload = function bootload(options, cb){
-  return nodefn.bindCallback(when.promise(function(resolve, reject) {
-
-    var serialPort;
-
-    if(!options){
-      return reject(new Error('Options error: no options'));
-    }
-
-    if(!options.path)
-    {
-      return reject(new Error('Options error: no path'));
-    }
-
-    if(!options.memory || !options.memory.data)
-    {
-      return reject(new Error('Options error: no program data'));
-    }
-
-    if(!options.board)
-    {
-      return reject(new Error('Options error: no board'));
-    }
-
-    function open(){
-      return when.promise(function(resolve, reject) {
-        serialPort = new com.SerialPort(options.path, { baudrate: 200 }, function(err){
-          if(err){ return reject(err); }
-          return resolve();
-        });
-      });
-    }
-
-    function setDtr(){
-      return when.promise(function(resolve, reject) {
-        serialPort.set({dtr: false}, function(err){
-          if(err){ return reject(err); }
-          return resolve();
-        });
-      });
-    }
-
-    function clrDtr(){
-      return when.promise(function(resolve, reject) {
-        serialPort.set({dtr: true}, function(err){
-          if(err){ return reject(err); }
-          return resolve();
-        });
-      });
-    }
-
-    function setBrk(){
-      return when.promise(function(resolve, reject) {
-        serialPort.write(new Buffer([0x00]), function(err){
-          if(err){ return reject(err); }
-          return resolve();
-        });
-      });
-    }
-
-    function clrBrk(){
-      return when.promise(function(resolve, reject) {
-        serialPort.update({baudRate: 9600}, function(err){
-          if(err){ return reject(err); }
-          return resolve();
-        });
-      });
-    }
-
-    function send(){
-      return bs2.bootload(serialPort, options.board, options.memory.data);
-    }
-
-    function close(){
-      return when.promise(function(resolve, reject) {
-
-        serialPort.on('error', function(err){
```

Comment:
Close isn't async so this is necessary where the connection might be opened again immediately as it wont be successfully closed until one of these return.